Sync browser url query with collection search parameters

dhmacs opened the pull request:
This PR closes #84 

Syncing search params with browser url query required to simplify the filter state logic. Moreover to improve management of side effects the search logic has been refactored using a state machine:

<img width="1059" alt="Screen Shot 2021-09-30 at 18 38 40" src="https://user-images.githubusercontent.com/4640135/135495954-0fca6add-7dc7-4208-9e36-09f12fd1f7d1.png">

> Tip: When using VSCode the machine can be visualized with [this extension](https://marketplace.visualstudio.com/items?itemName=mattpocock.xstate-vscode)

This change helped also to reduce re-renders and avoid the initial search when the collection is server side rendered. 
Since `@xstate/fsm` supports only a subset of the full XState library, some of the missing features that are required for this PR have been implemented under `lib/fsm-utils`. This way we get the benefit of modeling state with xstate without the bundle size cost of the full library.

> Note: the search error state has emerged as an edge case that we're currently not handling, but I decided to handle it in a separate PR

## QA

1. Visit the Vercel PR preview
2. Visit the parts collection page
3. Verify that when filters are used the url is updated
4. Verify that when searching for something the url is updated
5. Verify that when the pagination is used the url is updated
6. Verify that search parameters persist when the page is reloaded

deltuh-vee commented:
QA 🎬 
When filters are used the url is updated ✅ 
When searching for something the url is updated ✅ 
When the pagination is used the url is updated ✅ 
Search parameters persist when the page is reloaded ✅ 

deploy_block 🧩 
Selecting options with slashes or quotes seems to break the page(e.g. Screwdriver style > 1/4” Driver to 1/4” Socket).
![CleanShot 2021-10-01 at 11 12 19](https://user-images.githubusercontent.com/5289715/135667666-c7af9583-f19d-465a-a98b-1347a35d82c5.gif)

dhmacs commented:
@deltuh-vee Next monday I'll fix it 👍

dhmacs commented:
@deltuh-vee I've fixed it 👍 

> Note: I've also found out that the collection was not actually filtered by its handle, so I've fixed that too. This means that to find screwdriver styles you have to look at the tools collection

ardelato commented:
QA 👍 

- [x] Special characters (`/`, `,`, `-`,`.`,`<space>`, `&`) are all escaped properly

---

deploy_block❓ should the view mode also be saved?

https://user-images.githubusercontent.com/22064420/135885829-d4255f85-027d-479b-b15c-bb96089d7e6a.mp4


## Steps Taken:
1. Repeated the steps from https://github.com/iFixit/react-commerce/pull/94#issuecomment-932450770
2. I also confirmed I could not inject javascript into the url through the search field
![url attack good](https://user-images.githubusercontent.com/22064420/135886063-56af09a2-dc03-4238-83f0-a87190bddb24.png)

dhmacs commented:
@ardelato I thought about saving it as a local setting (local storage #95  ), so it's out of the scope of this PR. 
Do we want also the view to be synced to the url?

ardelato commented:
@dhmacs I don't think saving it as a url parameter is necessary. I'm okay with saving it as a local setting 👍 

un_deploy_block 👍

sterlinghirsh commented:
CR :zap:

dhmacs commented on features/collection/components/sections/FilterableProductsSection/FilterList/FacetFilter.tsx:
```diff
@@ -0,0 +1,90 @@
+      return options;
+   }, [facetOptions, filteredOptions, showAllValues, sortItems]);
+
+   const handleChange = React.useCallback(
+      (optionHandle: string) => {
+         toggle(optionHandle);
+      },
+      [toggle]
+   );
```
I think so.. It looks like we can use `toggle` directly 👍 Maybe I'll refactor this in a future PR
